### PR TITLE
added usability test cases for Operator/Domain interaction 

### PR DESCRIPTION
**Pull request by marinakog**
Unintall/Release Helm Chart | Deleting Operator Helm chart should not affect WebLogic domain Runtime functionality |  
-- | -- | --
Update the Operator Helm with New Domain NameSpace | The New Domain must be manged by the Operator |  
Update the Operator Helm  by removing a Domain NameSpace | The operator should no longer manage the Domain

### CODE REVIEW

**Comment by vanajamukkara** [comment]
@marinakog I marked the PR as WIP, you can change it to Ready to merge once you have tested and the code is reviewed. Let me know once the PR is ready for review.

**Comment by marinakog** [comment]
https://github.com/oracle/weblogic-kubernetes-operator/pull/943

**Comment by marinakog** [comment]
change is ready for review

**Comment by vanajamukkara** [comment]
A PR is in WIP until its verified on Jenkins and Wercker with clean runs and code review is complete.

File: integration-tests/src/test/java/oracle/kubernetes/operator/BaseTest.java [hunk 1]
```diff
@@ -489,6 +489,20 @@ public static ExecResult cleanup() throws Exception {
     return ExecCommand.exec(cmd);
   }
 
+  public static ExecResult cleanupDomainResources(String domainUid) throws Exception {
```
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/kubernetes/operator/BaseTest.java]: There is already one in TestUilts.deleteWeblogicDomainResources, you can use that.

File: integration-tests/src/test/java/oracle/kubernetes/operator/ITUsabilityOperatorHelmChart.java [hunk 2]
```diff
@@ -596,4 +581,127 @@ public void testCreateWithDefaultTargetDomainInstall() throws Exception {
...
+    Map<String, Object> operatorMap = TestUtils.createOperatorMap(number, true);
+    Operator operator = new Operator(operatorMap, RESTCertType.SELF_SIGNED);
+    operator.callHelmInstall();
+    Domain domain = null;
+    Domain domainnew = null;
+    boolean testCompletedSuccessfully = false;
+    logger.info("kubectl create namespace test" + (number + 1));
+    ExecCommand.exec("kubectl create namespace test" + (number + 1));
+    logger.info("create domain with UID : test" + number);
+    domain = TestUtils.createDomain(TestUtils.createDomainMap(number));
```
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/kubernetes/operator/ITUsabilityOperatorHelmChart.java]: This test method looks lengthy, can u move some of the code into private methods like createDomainAndVerify() or something like that..

File: integration-tests/src/test/java/oracle/kubernetes/operator/BaseTest.java (hunk 1 above)
Comment (marinakog) [reply integration-tests/src/test/java/oracle/kubernetes/operator/BaseTest.java]: updated and removed it from Base

File: integration-tests/src/test/java/oracle/kubernetes/operator/ITUsabilityOperatorHelmChart.java (hunk 2 above)
Comment (marinakog) [reply integration-tests/src/test/java/oracle/kubernetes/operator/ITUsabilityOperatorHelmChart.java]: changed to add private methods

File: integration-tests/README.md [hunk 3]
```diff
@@ -77,7 +77,8 @@ Basic Use Cases described above are verified in all the domain configurations. A
 | Two Operators using same External Https Port | create chart using same https rest port as already running first operator, verify that deployment fails with expected error |
 | Two Operators using same target domains namespace | create chart using target domains namespace as already running first operator, verify that deployment fails with expected error |
 | Operator Helm Chart using not preexisted target domains namespace | create chart using not preexisted target domains namespace as already running first operator, verify that deployment fails with expected error |
-
+| Operator Helm Chart add/delete target domains namespace ( domain1, domain2) | create chart , use upgrade to add/remove target domains ( domain1,domain2), verify that operator is able to manage added domain (domain2, not able to access the deleted one(domain1) |
```
Comment (markxnelson) [inline integration-tests/README.md]: no space after left parenthesis or before right parenthesis, no space before comma.

Comment (markxnelson) [reply integration-tests/README.md]: sentence does not make sense

File: integration-tests/src/test/java/oracle/kubernetes/operator/ITUsabilityOperatorHelmChart.java [hunk 4]
```diff
@@ -6,16 +6,9 @@
 
 import java.util.ArrayList;
 import java.util.Map;
-import oracle.kubernetes.operator.utils.ExecCommand;
-import oracle.kubernetes.operator.utils.ExecResult;
-import oracle.kubernetes.operator.utils.Operator;
+import oracle.kubernetes.operator.utils.*;
```
Comment (markxnelson) [inline integration-tests/src/test/java/oracle/kubernetes/operator/ITUsabilityOperatorHelmChart.java]: no wildcard imports

File: integration-tests/src/test/java/oracle/kubernetes/operator/ITUsabilityOperatorHelmChart.java [hunk 5]
```diff
@@ -6,16 +6,9 @@
...
 import oracle.kubernetes.operator.utils.Operator.RESTCertType;
-import oracle.kubernetes.operator.utils.TestUtils;
-import org.junit.AfterClass;
-import org.junit.Assume;
-import org.junit.BeforeClass;
-import org.junit.FixMethodOrder;
-import org.junit.Test;
+import org.junit.*;
```
Comment (markxnelson) [inline integration-tests/src/test/java/oracle/kubernetes/operator/ITUsabilityOperatorHelmChart.java]: no wildcard imports

File: integration-tests/src/test/java/oracle/kubernetes/operator/ITUsabilityOperatorHelmChart.java [hunk 6]
```diff
@@ -596,4 +581,130 @@ public void testCreateWithDefaultTargetDomainInstall() throws Exception {
...
+          (ArrayList<String>) (operator.getOperatorMap().get("domainNamespaces"));
+      targetDomainsNS.add("test" + (number + 1));
+      upgradeVerifyOperatorDomainNamespaces(operator, targetDomainsNS);
+      domainnew = createVerifyDomain(number + 1, operator);
+      logger.info("verify that old domain is managed by operator after upgrade");
+      operator.verifyDomainExists(domain.getDomainUid());
+      logger.info("Upgrade to remove first domain");
+      targetDomainsNS.remove("test" + (number));
+      upgradeVerifyOperatorDomainNamespaces(operator, targetDomainsNS);
+      Thread.sleep(60 * 1000);
```
Comment (markxnelson) [inline integration-tests/src/test/java/oracle/kubernetes/operator/ITUsabilityOperatorHelmChart.java]: i would prefer a loop to check for completion and then continue, so we do not add artificial and unnecessary delays to our test suite please.  we are trying to minimize the time taken to run tests.

File: integration-tests/src/test/java/oracle/kubernetes/operator/ITUsabilityOperatorHelmChart.java [hunk 7]
```diff
@@ -596,4 +581,130 @@ public void testCreateWithDefaultTargetDomainInstall() throws Exception {
...
+          "FAILURE: After Helm Upgrade for the domainNamespaces operator still able to manage old namespace ");
+    } catch (Exception ex) {
+      if (!ex.getMessage()
+          .contains("Response {\"status\":404,\"detail\":\"/operator/latest/domains/test" + number))
+        throw new RuntimeException(
+            "FAILURE: Exception does not report the expected error message " + ex.getMessage());
+
+      testCompletedSuccessfully = true;
+    } finally {
+      if (domain != null && !SMOKETEST && (JENKINS || testCompletedSuccessfully))
```
Comment (markxnelson) [inline integration-tests/src/test/java/oracle/kubernetes/operator/ITUsabilityOperatorHelmChart.java]: braces missing

File: integration-tests/src/test/java/oracle/kubernetes/operator/ITUsabilityOperatorHelmChart.java [hunk 8]
```diff
@@ -596,4 +581,130 @@ public void testCreateWithDefaultTargetDomainInstall() throws Exception {
...
+
+      if (operator != null) {
+        operator.destroy();
+      }
+      number++;
+    }
+
+    logger.info("SUCCESS - " + testMethodName);
+  }
+  /**
```
Comment (markxnelson) [inline integration-tests/src/test/java/oracle/kubernetes/operator/ITUsabilityOperatorHelmChart.java]: leave empty line after method

File: integration-tests/src/test/java/oracle/kubernetes/operator/ITUsabilityOperatorHelmChart.java [hunk 9]
```diff
@@ -596,4 +581,130 @@ public void testCreateWithDefaultTargetDomainInstall() throws Exception {
...
+    boolean testCompletedSuccessfully = false;
+    try {
+      Map<String, Object> operatorMap = TestUtils.createOperatorMap(number, true);
+      operator = new Operator(operatorMap, RESTCertType.SELF_SIGNED);
+      operator.callHelmInstall();
+      domain = createVerifyDomain(number, operator);
+      logger.info("Deleting operator to check that domain functionality is not effected");
+      operator.destroy();
+      operator = null;
+      Thread.sleep(60 * 1000);
```
Comment (markxnelson) [inline integration-tests/src/test/java/oracle/kubernetes/operator/ITUsabilityOperatorHelmChart.java]: don't introduce artificial delay

File: integration-tests/src/test/java/oracle/kubernetes/operator/ITUsabilityOperatorHelmChart.java [hunk 10]
```diff
@@ -596,4 +581,130 @@ public void testCreateWithDefaultTargetDomainInstall() throws Exception {
...
+      domain.testWlsLivenessProbe();
+      testCompletedSuccessfully = true;
+    } finally {
+      if (domain != null && !SMOKETEST && (JENKINS || testCompletedSuccessfully))
```
Comment (markxnelson) [inline integration-tests/src/test/java/oracle/kubernetes/operator/ITUsabilityOperatorHelmChart.java]: braces

File: integration-tests/src/test/java/oracle/kubernetes/operator/ITUsabilityOperatorHelmChart.java [hunk 11]
```diff
@@ -596,4 +581,130 @@ public void testCreateWithDefaultTargetDomainInstall() throws Exception {
...
+    domain.verifyDomainCreated();
+    testAdminT3Channel(domain);
+    TestUtils.renewK8sClusterLease(getProjectRoot(), getLeaseId());
+    logger.info("verify that domain is managed by operator");
+    operator.verifyDomainExists(domain.getDomainUid());
+    return domain;
+  }
+
+  private void upgradeVerifyOperatorDomainNamespaces(
+      Operator operator, ArrayList<String> targetNamespaces) throws Exception {
```
Comment (markxnelson) [inline integration-tests/src/test/java/oracle/kubernetes/operator/ITUsabilityOperatorHelmChart.java]: indentation

File: integration-tests/src/test/java/oracle/kubernetes/operator/utils/Operator.java [hunk 12]
```diff
@@ -324,6 +324,22 @@ public void callHelmUpgrade(String upgradeSet) throws Exception {
     logger.info("Command returned " + outputStr);
   }
 
+  public String getHelmValues() throws Exception {
```
Comment (markxnelson) [inline integration-tests/src/test/java/oracle/kubernetes/operator/utils/Operator.java]: you must update the copyright notice when you edit a file

File: integration-tests/README.md (hunk 3 above)
Comment (marinakog) [reply integration-tests/README.md]: updated

**Comment by marinakog** [comment]
all Usability tests passed in http://wls-jenkins.us.oracle.com/job/weblogic-kubernetes-operator-javatest/1332/consoleText, there are known failures for scaling in other integration tests

**Comment by vanajamukkara** [comment]
> 
> 
> all Usability tests passed in http://wls-jenkins.us.oracle.com/job/weblogic-kubernetes-operator-javatest/1332/consoleText, there are known failures for scaling in other integration tests

Looks like you have not synced to the latest develop branch. can you sync to the latest develop changes and re-run Jenkins?

File: integration-tests/src/test/java/oracle/kubernetes/operator/ITUsabilityOperatorHelmChart.java [hunk 13]
```diff
@@ -609,39 +620,28 @@ public void testAddRemoveDomainUpdateOperatorHC() throws Exception {
...
+      verifyOperatorDomainManagement(operator, domain, false);
+      verifyOperatorDomainManagement(operator, domainnew, true);
       testCompletedSuccessfully = true;
     } finally {
-      if (domain != null && !SMOKETEST && (JENKINS || testCompletedSuccessfully))
-        TestUtils.deleteWeblogicDomainResources(domain.getDomainUid());
-      if (domainnew != null && !SMOKETEST && (JENKINS || testCompletedSuccessfully))
-        domainnew.destroy();
-
+      if (domain != null) TestUtils.deleteWeblogicDomainResources(domain.getDomainUid());
```
Comment (markxnelson) [inline integration-tests/src/test/java/oracle/kubernetes/operator/ITUsabilityOperatorHelmChart.java]: braces missing

File: integration-tests/src/test/java/oracle/kubernetes/operator/ITUsabilityOperatorHelmChart.java [hunk 14]
```diff
@@ -666,22 +666,49 @@ public void testDeleteOperatorButNotDomain() throws Exception {
...
       operator.destroy();
       operator = null;
-      Thread.sleep(60 * 1000);
       domain.testWlsLivenessProbe();
       testCompletedSuccessfully = true;
     } finally {
-      if (domain != null && !SMOKETEST && (JENKINS || testCompletedSuccessfully))
-        TestUtils.deleteWeblogicDomainResources(domain.getDomainUid());
-
+      if (domain != null) TestUtils.deleteWeblogicDomainResources(domain.getDomainUid());
```
Comment (markxnelson) [inline integration-tests/src/test/java/oracle/kubernetes/operator/ITUsabilityOperatorHelmChart.java]: braces

File: integration-tests/src/test/java/oracle/kubernetes/operator/ITUsabilityOperatorHelmChart.java [hunk 15]
```diff
@@ -666,22 +666,49 @@ public void testDeleteOperatorButNotDomain() throws Exception {
...
+              .contains(
+                  "Response {\"status\":404,\"detail\":\"/operator/latest/domains/test" + number)) {
+          } else {
+            break;
+          }
+        }
+      }
+      if (i == maxIterations - 1) {
+        String errorMsg = "FAILURE: Operator can't access the domain " + domain.getDomainUid();
+        if (!isAccessible)
```
Comment (markxnelson) [inline integration-tests/src/test/java/oracle/kubernetes/operator/ITUsabilityOperatorHelmChart.java]: braces

File: src/integration-tests/apps/webservices/META-INF/application.xml [hunk 16]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (markxnelson) [inline src/integration-tests/apps/webservices/META-INF/application.xml]: copyright

File: src/integration-tests/apps/webservices/web/WEB-INF/web.xml [hunk 17]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (markxnelson) [inline src/integration-tests/apps/webservices/web/WEB-INF/web.xml]: copyright

File: integration-tests/src/test/java/oracle/kubernetes/operator/ITUsabilityOperatorHelmChart.java (hunk 13 above)
Comment (marinakog) [reply integration-tests/src/test/java/oracle/kubernetes/operator/ITUsabilityOperatorHelmChart.java]: so sorry, missed it , added braces

File: integration-tests/src/test/java/oracle/kubernetes/operator/ITUsabilityOperatorHelmChart.java (hunk 15 above)
Comment (marinakog) [reply integration-tests/src/test/java/oracle/kubernetes/operator/ITUsabilityOperatorHelmChart.java]: sorry, added braces

File: src/integration-tests/apps/webservices/web/WEB-INF/web.xml (hunk 17 above)
Comment (marinakog) [reply src/integration-tests/apps/webservices/web/WEB-INF/web.xml]: this file should not be in the PR, removed

File: src/integration-tests/apps/webservices/META-INF/application.xml (hunk 16 above)
Comment (marinakog) [reply src/integration-tests/apps/webservices/META-INF/application.xml]: this file should not be in the PR, removed

**Comment by marinakog** [comment]
http://wls-jenkins.us.oracle.com/job/weblogic-kubernetes-operator-javatest/1348

just one test fails same as develop

org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   ITOperator.testDomainOnPVUsingWDT:131->BaseTest.testWLDFScaling:456 ? Runtime ...
[INFO] 
[ERROR] Tests run: 32, Failures: 0, Errors: 1, Skipped: 0

File: integration-tests/src/test/java/oracle/kubernetes/operator/ITUsabilityOperatorHelmChart.java (hunk 11 above)
Comment (marinakog) [reply integration-tests/src/test/java/oracle/kubernetes/operator/ITUsabilityOperatorHelmChart.java]: mvn format (mvn com.coveo:fmt-maven-plugin:format)always put it in this format
 private void upgradeOperatorDomainNamespaces(
      Operator operator, ArrayList<String> targetNamespaces) throws Exception {
not sure what to do about it

File: integration-tests/README.md [hunk 18]
```diff
@@ -79,6 +79,16 @@ Basic Use Cases described above are verified in all the domain configurations. A
 | Operator Helm Chart using not preexisted target domains namespace | create chart using not preexisted target domains namespace as already running first operator, verify that deployment fails with expected error |
 | Operator Helm Chart add/delete target domain namespaces (domain1, domain2) | create operator helm chart managing domain1, use upgrade to add domain2. Verify that operator is able to manage added domain (domain2). Use helm upgrade to remove domain1, verify that operator not able to manage anymore the deleted one(domain1) |
 | Operator Helm Chart delete operator helm chart, leave domain running | create operator helm chart and start domain1, delete operator helm chart, verify domain1 is still functional |
+ 
+| Server Pods Restarted by modifying properties on the domain resource| Use Case |
+| --- | --- |
+| Server pods restarted by changing Env property | Verify admin and managed server pods being restarted by property change: "-Dweblogic.StdoutDebugEnabled=false" --> "-Dweblogic.StdoutDebugEnabled=true" |
```
Comment (markxnelson) [inline integration-tests/README.md]: can you please use back ticks not quotes for these?

Comment (marinakog) [reply integration-tests/README.md]: changed

**Pull request merged**